### PR TITLE
Add Range interface

### PR DESCRIPTION
**Pull request by onelxj**
`Range` interface gives the stream that iterates through the sequence of given ints.
Examples:
```
Range(10, 20)
| Map(Then([](int x){ return x + 5; }))
| Collect<std::vector<int>>
```

### CODE REVIEW

File: stout/range.h [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+////////////////////////////////////////////////////////////////////////
+
+} // namespace detail
+
+////////////////////////////////////////////////////////////////////////
+
+auto Range(int from, int to) {
+  return detail::_Range::Composable<int>{from, to};
+}
+
```
Comment (rjhuijsman) [inline stout/range.h:86]: Shall we also add a:
```c++
auto Range(int to) {
```
... Which assumes that `from` is `0`? That matches behavior on e.g. Python's `for i in range(10)`.

Comment (onelxj) [reply stout/range.h:86]: I guess we also can add a `step` as a third parameter:
`Range(0,10,2)` -> 0 2 4 6 8

Comment (rjhuijsman) [reply stout/range.h:86]: That's a good idea! Let's offer all three.

Comment (benh) [reply stout/range.h:86]: I like it!

File: stout/range.h [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+#include "stout/stream.h"
+
+////////////////////////////////////////////////////////////////////////
+
+namespace stout {
+namespace eventuals {
+
+////////////////////////////////////////////////////////////////////////
+
+namespace detail {
```
Comment (benh) [inline stout/range.h:14]: We're slowing moving away from using the `detail` namespace and just using the prefixed `_` instead, so please drop!

File: stout/range.h [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    template <typename Arg, typename K>
+    auto k(K k) && {
+      return Continuation<K, Arg>{std::move(k), from_, to_};
+    }
+
+    int from_;
+    int to_;
+  };
+};
+
```
Comment (benh) [inline stout/range.h:75]: Extra newline you can drop.

File: stout/range.h [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+    void Done() override {
+      k_.Ended();
+    }
+
+    K_ k_;
+    int to_;
+    int from_;
+  };
+
+  template <typename T>
```
Comment (benh) [inline stout/range.h:61]: Since you're not using this functionality right now (you're always using 'int') and it's really trivial to add later let's drop the template here.

File: BUILD.bazel [hunk 5]
```diff
@@ -41,6 +41,7 @@ cc_library(
         "stout/collect.h",
         "stout/filter.h",
         "stout/take.h",
+        "stout/range.h",
```
Comment (benh) [inline BUILD.bazel:44]: Sorry but you'll need to update this because we broke everything into multiple sub-libraries.

**Review by benh — CHANGES_REQUESTED** [review]

File: stout/range.h [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+        k_.Body(temp);
+      }
+    }
+
+    void Done() override {
+      k_.Ended();
+    }
+
+    K_ k_;
+    int to_;
```
Comment (rjhuijsman) [inline stout/range.h:56]: Looks like we can declare this `const`.

File: stout/range.h [hunk 7]
```diff
@@ -0,0 +1,93 @@
...
+
+struct _Range {
+  template <typename K_, typename Arg_>
+  struct Continuation : public detail::TypeErasedStream {
+    // NOTE: explicit constructor because inheriting 'TypeErasedStream'.
+    Continuation(K_ k, int from, int to, int step)
+      : k_(std::move(k)),
+        to_(to),
+        from_(std::min(from, to)),
+        step_(step) {}
```
Comment (rjhuijsman) [inline stout/range.h:22]: Small thing: let's keep the order of `from`, `to` and `step` consistent - same order in the constructor, parameter lists, and definition.

File: stout/range.h [hunk 8]
```diff
@@ -0,0 +1,93 @@
...
+    using ValueFrom = int;
+
+    template <typename Arg, typename K>
+    auto k(K k) && {
+      return Continuation<K, Arg>{std::move(k), from_, to_, step_};
+    }
+
+    int from_;
+    int to_;
+    int step_;
```
Comment (rjhuijsman) [inline stout/range.h:72]: Looks like we can declare these `const`?

File: test/range.cc [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+  EXPECT_THAT(*s, ElementsAre(-2, -1, 0, 1));
+}
+
+TEST(Range, DefaultFrom) {
+  auto s = Range(3)
+      | Collect<std::vector<int>>();
+
+  EXPECT_THAT(*s, ElementsAre(0, 1, 2));
+}
```
Comment (rjhuijsman) [inline test/range.cc:40]: Let's also have a version of this test that has a negative `to`.

File: test/range.cc [hunk 10]
```diff
@@ -0,0 +1,47 @@
...
+
+TEST(Range, SpecifiedStep) {
+  auto s = Range(0, 10, 2)
+      | Collect<std::vector<int>>();
+
+  EXPECT_THAT(*s, ElementsAre(0, 2, 4, 6, 8));
+}
```
Comment (rjhuijsman) [inline test/range.cc:47]: Let's also have a version of this test that has a negative `step`.

**Review by rjhuijsman — CHANGES_REQUESTED** [review]
